packages/astro/snowpack-plugin-jsx.cjs

```diff
@@ -0,0 +1,155 @@
+const esbuild = require('esbuild');
+const colors = require('kleur/colors');
+const { logger } = require('snowpack');
+const path = require('path');
+const { promises: fs } = require('fs');
+
+const babel = require('@babel/core')
+const eslexer = require('es-module-lexer');
+
+/**
+ * @typedef {Object} PluginOptions - creates a new type named 'SpecialType'
+ * @prop {import('./src/config_manager').ConfigManager} configManager
+ * @prop {'development' | 'production'} mode
+ */
+
+/**
+ * Returns esbuild loader for a given file
+ * @param filePath {string}
+ * @returns {import('esbuild').Loader}
+ */
+function getLoader(fileExt) {
+  /** @type {any} */
+  return fileExt.substr(1);
+}
+
+/**
+ * @type {import('snowpack').SnowpackPluginFactory<PluginOptions>}
+ */
+module.exports = function jsxPlugin(config, options = {}) {
+  const {
+    configManager
+  } = options;
+
+  let didInit = false;
+  return {
+    name: '@astrojs/snowpack-plugin-jsx',
+    resolve: {
+      input: ['.jsx', '.tsx'],
+      output: ['.js'],
+    },
+    async load({ filePath, fileExt, ...transformContext }) {
+      if (!didInit) {
+        await eslexer.init;
+        didInit = true;
+      }
+
+      const contents = await fs.readFile(filePath, 'utf8');
+      const loader = getLoader(fileExt);
+
+      const { code, warnings } = await esbuild.transform(contents, {
+        loader,
+        jsx: 'preserve',
+        sourcefile: filePath,
+        sourcemap: config.buildOptions.sourcemap ? 'inline' : undefined,
+        charset: 'utf8',
+        sourcesContent: config.mode !== 'production',
+      });
+      for (const warning of warnings) {
+        logger.error(`${colors.bold('!')} ${filePath}
+  ${warning.text}`);
+      }
+
+      let renderers = await configManager.getRenderers();
+      const importSources = new Set(renderers.map(({ jsxImportSource }) => jsxImportSource).filter(i => i));
+      const getRenderer = (importSource) => renderers.find(({ jsxImportSource }) => jsxImportSource === importSource);
+      const getTransformOptions = async (importSource) => {
+        const { name } = getRenderer(importSource);
+        const { default: renderer } = await import(name);
+        return renderer.jsxTransformOptions(transformContext);
+      }
+
+      // If we only have a single renderer, we can skip a bunch of work!
+      if (importSources.size === 1) {
+        const result = transform(code, filePath, await getTransformOptions(Array.from(importSources)[0]))
+
+        return {
+          '.js': {
+            code: result.code || ''
+          },
+        };
+      }
+
+      // we need valid JS to scan for imports
+      // so let's just use `h` and `Fragment` as placeholders
+      const { code: codeToScan } = await esbuild.transform(code, {
```

Comment:
Unfortunately, if we have multiple renderers, we need to scan their imports. This means one more pass through `esbuild` to get valid JS (no JSX). The result doesn't need to run, it will be discarded.

Reply:
I'm really -1 on a build failing when imports change, basically the user story of:

1. I refactor a file of React components to remove the need for `useState`
2. I clean up my imports, removing `import {useState} from 'react';`
3. my build fails because the imports changed.

Can we just be more explicit, and always use the first matched renderer in your config file _unless_ the `jsxImportSource` pragma comment is used?

I know we _can_ be smart than this with the imports, but I think it leads to more surprises (the bad kind) in practice.

Reply:
@FredKSchott that refactor hazard has existed as long as JSX has though. I think requiring people add a non-standard comment to every file is asking quite a lot. The mental model of a JSX expression being syntactic sugar for a function call makes sense. Now making it an external config value feels like a regression to me.

Reply:
@matthewp Agreed, but I would point out that the comment is borrowed from [`@babel/plugin-transform-react-jsx`](https://babeljs.io/docs/en/babel-plugin-transform-react-jsx#customizing-the-automatic-runtime-import).

@FredKSchott I'm wary of having the order of the renderers in the config array impact the behavior in an unintuitive way. I've noticed that trip a lot of people up with other tools. I'm also wary of adding another property like `fallbackRenderer` that only applies for JSX. Can we solve this with a good error message?


```
Unable to resolve JSX transformer! With more than one renderer enabled, you should use a pragma comment or include an import.
Add `/* jsxImportSource: preact */` or `import { h } from "preact"`
```

Reply:
This just feels a lot like we're optimizing for the 1% use case at the expense of the 99% use case. If most users are only ever using one framework, they're going to see this warning even though they only have one JSX renderer connected.

How about:
- If the user only has 1 JSX renderer defined in config, default to that
- If the user has 2+ JSX renderers defined in config, show the warning from the comment above

Reply:
Also if we're word-smithing the warning, I'd say:

```
Unable to detect JSX framework! Add `/* jsxImportSource: preact */` to the file to fix.
```

Reply:
A user that only has one framework won't run into this issue, the code always uses the first import source, unless I'm misreading this: https://github.com/snowpackjs/astro/pull/588/files#diff-4326d59c6d077cb5cc64b38132f64e52a47473932d71225703a0fae2fe6b59adR84

I also disagree with the notion that framework swapping is going to be rare. I think that's y

Reply:
A user with one framework won't run into the import scanning because it's already using the first importSource in that case, unless I'm misunderstanding this code: https://github.com/snowpackjs/astro/pull/588/files#diff-4326d59c6d077cb5cc64b38132f64e52a47473932d71225703a0fae2fe6b59adR84

Reply:
doh, I missed that and I even commented on that line! If that's already what you're proposing @natemoo-re then I'm on board 👍

Reply:
That's correct! This only ever warns if a user is using multiple JSX-aware renderers. I also improved the warning to suggest the user's first configured renderer automatically.